src/mapping/query/function/format_url.rs

```diff
@@ -0,0 +1,235 @@
+        let query = arguments.required("value")?;
+
+        Ok(Self { query })
+    }
+}
+
+fn components_to_url(components: BTreeMap<String, Value>) -> Result<Url> {
+    let scheme = components
+        .get(SCHEME)
+        .map(|v| string(SCHEME, v))
+        .transpose()?
+        .ok_or_else(|| missing(SCHEME))?;
+
+    let username = components
+        .get(USERNAME)
+        .map(|v| string(USERNAME, v))
+        .transpose()?
+        .unwrap_or_else(|| "".to_owned());
+
+    let password = components
+        .get(PASSWORD)
+        .and_then(Value::as_option)
+        .map(|v| string(PASSWORD, v))
+        .transpose()?;
+
+    let host = components
+        .get(HOST)
+        .and_then(Value::as_option)
+        .map(|v| string(HOST, v))
+        .transpose()?;
+
+    let port = components
+        .get(PORT)
+        .and_then(Value::as_option)
+        .map(|v| {
+            v.as_i64()
+                .map(|i| i as u16)
+                .ok_or_else(|| invalid(PORT, "map", v.kind()))
+        })
+        .transpose()?;
+
+    let path = components
+        .get(PATH)
+        .map(|v| string(PATH, v))
+        .transpose()?
+        .unwrap_or_else(|| "".to_owned());
+
+    let fragment = components
+        .get(FRAGMENT)
+        .and_then(Value::as_option)
+        .map(|v| string(FRAGMENT, v))
+        .transpose()?;
+
+    let query = components
+        .get(QUERY)
+        .map(|v| v.as_map().ok_or_else(|| invalid(QUERY, "map", v.kind())))
+        .transpose()?
+        .map(|v| {
+            v.iter()
+                .map(|(k, v)| (k, v.to_string_lossy()))
+                .collect::<Vec<_>>()
+        })
+        .unwrap_or_default();
+
+    // This is a really awkward way to create a valid URL from components, but
+    // it doesn't appear the `Url` API offers any better way to handle this.
+    let mut url = Url::parse("https://vector.dev").expect("valid");
+
+    url.set_scheme(&scheme).map_err(|_| unknown(SCHEME))?;
+    url.set_host(host.as_deref()).map_err(|_| unknown(HOST))?;
+    url.set_username(&username).map_err(|_| unknown(USERNAME))?;
+    url.set_password(password.as_deref())
+        .map_err(|_| unknown(PASSWORD))?;
+    url.set_path(&path);
+    url.set_port(port).map_err(|_| unknown(PORT))?;
+    url.set_fragment(fragment.as_deref());
+
+    if !query.is_empty() {
+        url.query_pairs_mut().extend_pairs(query);
+    }
+
+    Ok(url)
+}
+
+fn string(component: &'static str, value: &Value) -> Result<String> {
+    value
+        .as_bytes2()
+        .ok_or_else(|| invalid(component, "string", value.kind()))
+        .map(|b| String::from_utf8_lossy(&b).into_owned())
+}
+
+#[inline]
+fn missing(s: &'static str) -> String {
+    format!("missing url component '{}'", s)
+}
+
+#[inline]
+fn invalid(component: &'static str, expected: &str, got: &str) -> String {
+    format!(
+        "invalid type for url component '{}' (expected {}, got {})",
+        component, expected, got,
+    )
+}
```

Comment:
I suspect at some point we'll want a more generic approach to deserialize a `Value::Map`, which we can use throughout remap functions dealing with maps, so that we don't have to re-do parsing and error handling, but I figured I'd do it manually for now, until we have a need to extract things out.

Reply:
This work seems to be about half of serde's deserialize operation, going from an intermediate form instead of a fully serialized form, into a struct. I wonder if you couldn't use some of the serde framework to do this, treating a `Value::Map` as the serialized form.